File: leetcode/67-add-binary.py
# https://leetcode.com/problems/add-binary/

class Solution:
    def addBinary(self, a: str, b: str) -> str:
        total_sum = ''
        carry = 0
        a = a[::-1]
        b = b[::-1]
        for i in range(max(len(a), len(b))):
            a_bit = int(a[i]) if i < len(a) else 0
            b_bit = int(b[i]) if i < len(b) else 0
            bit_sum = a_bit + b_bit + carry
            total_sum = str(bit_sum % 2) + total_sum
            carry = 1 if bit_sum > 1 else 0 # (bit_sum // 2) work also

        if carry == 1:
            total_sum = '1' + total_sum
        return total_sum
    

    
    # this also work...
    def add_binary(self, a: str, b: str) -> str:
        total_sum = int(a, 2) + int(b, 2)
        return bin(total_sum)[2:]
        

File: leetcode/67-add-binary.js
// https://leetcode.com/problems/add-binary/

/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
var addBinary = function(a, b) {
  
  const totalSum = [];

  a = a.split('').reverse().join('');
  b = b.split('').reverse().join('');

  let carray = 0;

  for (let i = 0; i < Math.max(a.length, b.length); i++) {
    
    const bitSum = carray + ( a.length > i ? Number(a[i]) : 0) + ( b.length > i ? Number(b[i]) : 0);

    switch(bitSum) {
      case 0: {
        totalSum.push('0');
        carray = 0;
        break;
      }
      case 1: {
        totalSum.push('1');
        carray = 0;
        break;
      }
      case 2: {
        totalSum.push('0');
        carray = 1;
        break;
      }
      case 3: {
        totalSum.push('1');
        carray = 1;
        break;
      }
    }

    // also work instead of the switch
    // if (bitSum === 0 || bitSum === 2) totalSum.push('0');
    // else totalSum.push('1'); // bitSum === 1 || bitSum === 3
    // if (bitSum >= 2) carray = 1;
    // else carray = 0;

    // also work instead of the switch
    // totalSum.push(String(bitSum % 2));
    // carray = Math.floor(bitSum / 2);
  }

  if (carray === 1) {
    totalSum.push('1');
  }
  
  return totalSum.reverse().join('');
};

const sum = addBinary('1010', '1011');
console.log(sum);